File: 链表/19.删除链表的倒数第N个结点.py

# 记得 dummy 虚拟节点的使用 处理难应付的边界情况
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        dummy = ListNode()
        dummy.next = head
        # 实际删除要找到要删除节点的前一个节点，所以end要多走一步，二者距离要增加1
        f_pre_del, f_end = dummy, head
        # 在没链表的长度内用f_end拉开距离
        while f_end and n:
            f_end = f_end.next
            n -= 1
        # 二者同时走，距离保持不变，直到f_end到结尾
        while f_end:
            f_pre_del = f_pre_del.next
            f_end = f_end.next
        f_pre_del.next = f_pre_del.next.next
        return dummy.next

File: 链表/19.删除链表的倒数第N个结点.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} n
 * @return {ListNode}
 */
var removeNthFromEnd = function (head, n) {
  const dummy = new ListNode(0, head);
  let pre = dummy;  // cur和pre都从dummy开始保证中间相差n个节点
  while (n) {
    pre = pre.next;
    n--;
  }

  let before = new ListNode(0, dummy);
  let after = dummy.next;
  while (pre) {
    pre = pre.next;
    before = before.next;
    after = after.next;
  }
  before.next = after;
  return dummy.next;
};


//? 剑指offer 22 获取倒数第k个节点
var _removeNthFromEnd = function (head, n) {
  const dummy = new ListNode(0, head);
  let cur = dummy, p = dummy;
  while (n) {
    p = p.next;
    n--;
  }
  while (p) {
    cur = cur.next;
    p = p.next;
  }
  return cur.val;
};
